salt/metalk8s/volumes/cleanup.sls

```diff
@@ -0,0 +1,56 @@
+# Clean already attached loop devices before applying highstate, to avoid
+# creating duplicates for a same sparse file.
+# This state can cleanup either "manually", for upgrade, or by disabling the
+# systemd units, for downgrade.
+{%- set target_version = pillar.metalk8s.nodes[grains.id].version %}
+
+include:
+  - .installed
+
+{%- if '_errors' in pillar.metalk8s.volumes %}
+
+Cannot proceed with volume cleanup (pillar errors):
+  test.fail_without_changes:
+    - comment: 'Errors: {{ pillar.metalk8s.volumes._errors | join (", ") }}'
+
+{%- else %}
+  {%- set sparse_volumes = pillar.metalk8s.volumes.values()
+                         | selectattr('spec.sparseLoopDevice', 'defined')
+                         | list %}
+
+  {%- if not sparse_volumes %}
+
+Nothing to cleanup:
+  test.succeed_without_changes: []
+
+  {%- else %}
+    {%- for volume in sparse_volumes %}
+      {%- set volume_name = volume.metadata.name %}
+      {%- set volume_id = volume.metadata.uid %}
+      {%- if salt.pkg.version_cmp(target_version, '2.7.0') >= 0 %}
+
+Cleanup loop device for Volume {{ volume_name }}:
+  cmd.run:
+    - name: /usr/local/libexec/metalk8s-sparse-volume-cleanup "{{ volume_id }}"
+    # Only cleanup if the systemd unit doesn't exist yet (this command exits
+    # with retcode 3 if the service is dead, 4 if the template does not exist)
+    - unless: systemctl status metalk8s-sparse-volume@{{ volume_id }}
+    - require:
+      - test: Ensure Python 3 is available
+      - file: Install clean-up script
+
+      {%- else %} {# target_version < 2.7.0 #}
+
+Disable systemd unit for Volume {{ volume_name }}:
+  service.dead:
+    - name: metalk8s-sparse-volume@{{ volume_id }}
+    - enable: false
+    - require:
+      - test: Ensure Python 3 is available
+      - file: Set up systemd template unit for sparse loop device provisioning
+      - file: Install clean-up script
+
+      {%- endif %}
```

Comment:
At the very end of the for loop we may want to remove the systemd unit file, so that nobody can manually enable back the service, no ?

Reply:
Maybe, but I'm not sure why someone would want to enable this service manually unless they know what they're doing. And not sure what the issue would be, except that they may get duplicate loop devices (but then again, they would know, no?).

Reply:
Hmm I don't know, 
I'm a user, I have some trouble on my platform, I check for "metalk8s" system services I see this one stopped, I start it :cry: :smile: 
Anyway it will happen only if a user do a downgrade and try to debug itself the cluster after this so ...I agree it's not critical and can be handled later

Reply:
Since I'm reworking this state, I'll add the cleanup of this unit template.
But I'm not sure how you can find the disabled+stopped services:

```
$ # output after a downgrade from this branch to 2.6.0
$ systemctl list-units -a | grep metalk8s
  srv-scality-metalk8s\x2d2.6.0.mount          loaded    active   mounted   /srv/scality/metalk8s-2.6.0
  srv-scality-metalk8s\x2d2.7.0\x2ddev.mount   loaded    active   mounted   /srv/scality/metalk8s-2.7.0-dev
  var-lib-metalk8s-storage-sparse.mount        loaded    active   mounted   /var/lib/metalk8s/storage/sparse
  system-metalk8s\x2dsparse\x2dvolume.slice    loaded    active   active    system-metalk8s\x2dsparse\x2dvolume.slice
```

Reply:
Hmm ok you right, you can have it with "systemctl list-unit-files" but yes ok you do not have the service, anyway would be cleaner to remove the unit file and it does not cost much

Reply:
Done